### DISCUSSION TITLE
Concurrency Control Clarification Question

### DISCUSSION THREAD

**Discussion by Alfonso-Martello** [iTwin/itwinjs-core#6115] [Q&A] [answered]
Hello,

I am helping a product team upgrade from iTwin 2.x to iTwin 3.x. In 2.x, the concurrency “mode” was determined by the application. In 3.x, the concurrency “mode” is determined at iModel creation time (as per the [changelogs]( https://www.itwinjs.org/changehistory/3.0.0/#concurrency-control)). The product team’s codebase appears to expect a pessimistic “mode”. However, does enforcing the concurrency “mode” matter for applications (especially since this is not explicitly possible)?

For example, let’s say I am performing an operation that requires a lock and I call [`acquireLocks`]( https://www.itwinjs.org/reference/core-backend/imodels/lockcontrol/acquirelocks/). If the iModel was created with the pessimistic “mode”, then it will acquire any needed locks. If the iModel was instead created with the optimistic “mode”, then it will not acquire any locks, and any changes will be merged (and this merging is not something the application itself needs to handle, it is handled by the platform code)?

**Comment by kabentley** [comment, marked as answer]
You can assume all iModels will be created with locking mode enabled - that's the default. It never made any sense for that to be a runtime application choice since it defeats the very purpose of locks if someone can "opt out" of them.

But no, using "no locks" mode will not magically merge changes from multiple users. As in Git, conflict resolution will require user (and likely application) intervention.
